src/Core/AdminConsole/OrganizationFeatures/OrganizationUsers/RevokeNonCompliantOrganizationUserCommand.cs

```diff
@@ -0,0 +1,115 @@
+using Bit.Core.Enums;
+using Bit.Core.Models.Commands;
+using Bit.Core.Models.Data.Organizations.OrganizationUsers;
+using Bit.Core.Repositories;
+using Bit.Core.Services;
+
+namespace Bit.Core.AdminConsole.OrganizationFeatures.OrganizationUsers;
+
+public class RevokeNonCompliantOrganizationUserCommand(IOrganizationUserRepository organizationUserRepository,
+    IEventService eventService,
+    IHasConfirmedOwnersExceptQuery confirmedOwnersExceptQuery,
+    TimeProvider timeProvider) : IRevokeNonCompliantOrganizationUserCommand
+{
+    public const string ErrorCannotRevokeSelf = "You cannot revoke yourself.";
+    public const string ErrorOnlyOwnersCanRevokeOtherOwners = "Only owners can revoke other owners.";
+    public const string ErrorUserAlreadyRevoked = "User is already revoked.";
+    public const string ErrorOrgMustHaveAtLeastOneOwner = "Organization must have at least one confirmed owner.";
+    public const string ErrorInvalidUsers = "Invalid users.";
+    public const string ErrorRequestedByWasNotValid = "Action was performed by an unexpected type.";
+
+    public async Task<CommandResult> RevokeNonCompliantOrganizationUsersAsync(RevokeOrganizationUsersRequest request)
+    {
+        var validationResult = await ValidateAsync(request);
+
+        if (validationResult.HasErrors)
+        {
+            return validationResult;
+        }
+
+        await organizationUserRepository.SetOrganizationUsersStatusAsync(request.OrganizationUsers.Select(x => x.Id),
+            OrganizationUserStatusType.Revoked);
+
+        if (request.ActionPerformedBy.GetType() == typeof(StandardUser))
+        {
+            await eventService.LogOrganizationUserEventsAsync(
+                request.OrganizationUsers.Select(x => GetRevokedUserEventTuple(x, timeProvider.GetUtcNow())));
+        }
+        else if (request.ActionPerformedBy is SystemUser { SystemUserType: not null } loggableSystem)
+        {
+            await eventService.LogOrganizationUserEventsAsync(
+                request.OrganizationUsers.Select(x =>
+                    GetRevokedUserEventBySystemUserTuple(x, loggableSystem.SystemUserType.Value,
+                        timeProvider.GetUtcNow())));
+        }
+
+        return validationResult;
+    }
+
+    private static (OrganizationUserUserDetails organizationUser, EventType eventType, DateTime? time) GetRevokedUserEventTuple(
+        OrganizationUserUserDetails organizationUser, DateTimeOffset dateTimeOffset) => new(organizationUser,
+        EventType.OrganizationUser_Revoked, dateTimeOffset.UtcDateTime);
+
+    private static (OrganizationUserUserDetails organizationUser, EventType eventType, EventSystemUser eventSystemUser, DateTime? time) GetRevokedUserEventBySystemUserTuple(
+        OrganizationUserUserDetails organizationUser, EventSystemUser systemUser, DateTimeOffset dateTimeOffset) => new(organizationUser,
+        EventType.OrganizationUser_Revoked, systemUser, dateTimeOffset.UtcDateTime);
```

Comment:
This seems like a lot of boilerplate when you could just pass in a tuple directly. Is it doing much for you?

Reply:
Idk. It removes noise from where the actual action is happening. I'd rather have simple method signature in the select than `new ValueTuple<OrganizationUserUserDetails, EventType, DateTime?>(organizationUser,
            EventType.OrganizationUser_Revoked, dateTimeOffset.UtcDateTime);`.

Reply:
Oh, it's explicitly a `ValueTuple` rather than just a (named) tuple. Yep then I agree that's cumbersome otherwise.